### PR TITLE
Change setting up temporary directories in `astropy` test runner

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

`astropy` v7.0.0 has a bug that causes the `XDG_CACHE_HOME` environment variable to have no effect. Instead `XDG_CONFIG_HOME` controls both the cache and configuration directories. #17514 would restore the correct pre-v7.0.0 behavior, but it failed the double test CI job. That job uses the `astropy` test runner instead of calling `pytest` directly, and our test runner sets up temporary directories with the `set_temp_cache` and `set_temp_config` context managers, which have priority over the environment variables. The failure in #17514 is therefore specific to our test runner and it is not revealing any mistakes in the regression test or in the relevant code. It would be possible to avoid this problem by complicating the regression test, but that is not a robust solution and it is not needed at all if `pytest` is called directly. It is our test runner that should be edited to avoid using the aforementioned context managers. Our `pytest` configuration already uses the environment variables instead of the context managers, so this change also makes our test runner configuration more similar to that.

I don't see any reason why the test runner should set up a temporary cache directory because we configure `pytest` to do that anyways (it actually creates more than one). It is also not obvious to me why our runner should set up a temporary configuration directory, but a code comment (by @astrofrog) mentions timing concerns. I chose to trust the comment, so I replaced rather than removed `set_temp_config`.

This pull request must be backported because otherwise backporting #17514 will cause double test CI job failures in the `v7.0.x` branch.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/tests/runner.py [hunk 1]
```diff
@@ -235,23 +235,25 @@ def run_tests(self, **kwargs):
...
+        # but it's also harmless
+        orig_xdg_config = os.environ.get("XDG_CONFIG_HOME")
+        with tempfile.TemporaryDirectory("astropy_config") as astropy_config:
+            Path(astropy_config, "astropy").mkdir()
+            os.environ["XDG_CONFIG_HOME"] = astropy_config
+            try:
                 return pytest.main(args=args, plugins=plugins)
+            finally:
+                if orig_xdg_config is None:
+                    os.environ.pop("XDG_CONFIG_HOME", None)
```
Comment (neutrinoceros) [inline astropy/tests/runner.py:254]: I did a double take on the second argument. I now agree this is the correct thing to do: even if this environment variable is *supposed* to be defined at this point, we cannot guarantee it is still there, and we certainly do no want to raise an exception within a `finally` clause.

**Review by neutrinoceros — APPROVED** [review]
LGTM. Thanks a lot for the comprehensive writeup ! even being involved in previous discussions it was very helpful for me to get back into this specific context.

**Comment by pllim** [comment]
Thanks! But seems like I didn't get to this soon enough.

> I don't see any reason why the test runner should set up a temporary cache directory because we configure pytest to do that anyways (it actually creates more than one)

Where is that done; can you please link?

I think the original reason was to ensure the second run does not accidentally read in cached stuff from first run, but @astrofrog can correct me if I am wrong. `__pycache__` is not the same as `.astropy/cache` if that is what you are referring to.

**Comment by eerovaher** [comment]
https://github.com/astropy/astropy/blob/abc83b4e661d3abf720dfe4c68f04ae48c220258/conftest.py#L69-L76

https://github.com/astropy/astropy/blob/abc83b4e661d3abf720dfe4c68f04ae48c220258/astropy/conftest.py#L87-L99

https://github.com/astropy/astropy/blob/abc83b4e661d3abf720dfe4c68f04ae48c220258/docs/conftest.py#L14-L21

And if the `astropy` test runner is used then that will create one more configuration directory.

It is simple to prevent our test runner from creating a configuration directory by invoking `pytest` directly. It is also simple to avoid executing `astropy/conftest.py` by running `pytest docs` and it is equally simple to avoid executing `docs/conftest.py` by running `pytest astropy`, but I have not managed to run the tests without the creation of at least two configuration and two cache directories.

From what I can tell having `conftest.py` create the directories should be enough. I don't really understand why our test runner should also create a configuration directory, but a comment mentions timing concerns, which can be tricky to investigate, so I just trusted the comment instead of looking into this properly.

**Comment by pllim** [comment]
Ah, right... `pytest.main` does respect `conftest.py`, theoretically. The test runner existed back when pytest was a *single file* and was bundled with astropy itself (shocking, isn't it?). So it is very possible that we patched different things over time to the point of accidentally duplicating the `mkdtemp` stuff.

Is there a way to check if `conftest.py` (the copy that test runner actually uses in `astropy.test()` which is probably the one that is `astropy/astropy/conftest.py`) is shipped in astropy wheel? The only reason I can think of when we have to repeat the `mkdtemp` in runner code is when user does not have access to that `conftest.py` when calling `astropy.test()`. I don't think the [wheel building log](https://github.com/astropy/astropy/actions/runs/12269442633/job/34233016572) tells me what files are included.

**Comment by neutrinoceros** [comment]
> Is there a way to check if conftest.py (the copy that test runner actually uses in astropy.test() which is probably the one that is astropy/astropy/conftest.py) is shipped in astropy wheel? 

Yes: `.whl` are actually just `.zip` in disguise, so one can just download any wheel from https://pypi.org/simple/astropy/ and `unzip` it to see what's inside. 
For instance
```
$ cd /tmp/
$ wget https://files.pythonhosted.org/packages/fa/3a/cf06361685f72b50b3fb69247b95b66a6d2edb1b1947e6403c168f01eecc/astropy-7.0.0-cp313-cp313-win_amd64.whl#sha256=7ecaaeeb017cceb02512d5505c43e162b3e26d951cece2b021e786f4c0af17a7
$ unzip unzip astropy-7.0.0-cp313-cp313-win_amd64.whl
$ ls astropy | grep conftest.py
conftest.py
```
so yes, it's in there (I checked that it is indeed identical to `astropy/astropy/conftest.py`)

**Pull request merged**